packages/doenetml-worker/lib-doenetml-core/src/dast/ref_resolve/root_names.rs

```diff
@@ -0,0 +1,269 @@
+    }
+
+    // The final root name for each node is the "simplest" of the reachable names where "simplest" is defined by
+    // the fewest number of pieces (indices or sub-names) used to construct the name.
+    // Ties are broken by the fewest number of indices used, followed by the shortest overall name.
+    reachable_names
+        .into_iter()
+        .map(|possible_names| {
+            possible_names.unwrap().into_iter().reduce(|acc, name| {
+                let index_count_name = name.chars().filter(|c| *c == ':').count();
+                let index_count_acc = acc.chars().filter(|c| *c == ':').count();
+                let separator_count_name =
+                    index_count_name + name.chars().filter(|c| *c == '.').count();
+                let separator_count_acc =
+                    index_count_acc + acc.chars().filter(|c| *c == '.').count();
+
+                if separator_count_name < separator_count_acc {
+                    name
+                } else if separator_count_name > separator_count_acc {
+                    acc
+                } else if index_count_name < index_count_acc {
+                    name
+                } else if index_count_name > index_count_acc {
+                    acc
+                } else if name.len() < acc.len() {
+                    name
+                } else {
+                    acc
+                }
+            })
+        })
+        .collect::<Vec<_>>()
+}
+
+// Given that `resolutions_to_node` gives all the resolutions of single names and indices,
+// concatenate all the combinations of name and index resolutions that form a path from the root to `node_idx`.
+// If `reachable_names` has any entries where the reachable names have already been computed,
+// use those to shortcut the algorithm
+fn concatenate_reachable_names(
+    node_idx: usize,
+    mut reachable_names: Vec<Option<NamesReachableFromRoot>>,
+    resolutions_to_node: &[ResolutionsToNode],
+) -> Vec<Option<NamesReachableFromRoot>> {
+    // if `reachable_names` for `node_idx` has already been computed in the course of a previous calculation,
+    // there is nothing to do.
+    if reachable_names[node_idx].is_some() {
+        return reachable_names;
+    }
+
+    let mut names = Vec::new();
+
+    let resolutions = &resolutions_to_node[node_idx];
+
+    if let Some(ResolutionViaName { name, origins }) = &resolutions.by_name {
+        // Since `node_idx` can be reached via `name` from `origins`,
+        // determine how one can reach each origin from root and concatenate `name` with the result
+        for origin in origins.iter() {
+            match origin {
+                // node can be directly reached from root via `name`
+                NodeOrRoot::Root => names.push(name.to_string()),
+                // since `origin` is not the root, we need to recurse
+                NodeOrRoot::Node(dest_idx) => {
+                    reachable_names = concatenate_reachable_names(
```

Comment:
This recursion happening from multiple places makes the code hard to follow. Can this process be done with a simple iterator instead?